Enable vara container install

boehmseb opened the pull request:
Resolves se-passau/VaRA#718
The dependency declarations in this PR will also be useful for se-passau/VaRA#626

boehmseb committed:
Enable vara container install

boehmseb commented on varats/varats/tools/research_tools/vara.py:
```diff
@@ -256,14 +264,27 @@ def verify_install(self, install_location: Path) -> bool:
+        container_vara_dir = containers.CONTAINER_VARATS_ROOT / "tools/VaRA"
+        layers.run(*(self.DEPENDENCIES.get_install_command(distro).split(" ")))
+        layers.copy_([str(self.install_location())], str(container_vara_dir))
+        layers.env(BB_ENV=f"{{PATH: [{str(container_vara_dir / 'bin')}]}}")
```
If we do it like this, we can set `BB_ENV` only at one place, since every call to `layers.env()` overwrites existing env vars in the container. What's your opinion on this @vulder

vulder replied:
Hmm, one location to define it seems fine to me. But maybe we should add code that checks what is defined in the current BB env not that we miss something. Not related to the path by maybe something like this.
BB_ENV = { 'enable_instrumentation': "FooOn"}

... inside the container ... a binary could look for the env var to enable/disable instrumentation code

boehmseb replied:
Well, that's actually a second point you are talking about here.
But I guess this will become a longer discussion we should do in person.

boehmseb replied:
Should be fixed once and for all by the new `BaseImageCreationContext` class.

vulder commented on varats/varats/tools/research_tools/research_tool.py:
```diff
@@ -29,6 +30,38 @@
+        Returns:
+            the command how the dependencies can be installed
+        """
+        return _install_commands[distro] + " " + " ".join(
+            self.__dependencies[distro]
+        )
```
Maybe use a f string here

vulder commented on varats/varats/tools/research_tools/research_tool.py:
```diff
@@ -29,6 +30,38 @@
+
+    def __init__(self, dependencies: tp.Dict[Distro, tp.List[str]]):
+        self.__dependencies = dependencies
+
```
Does it make sense to add a `add` method for a new dependency?

boehmseb replied:
I don't see any use case for this right now. Currently, we only want to use them for research tools, and there all dependencies are known statically.
But this class will get additional functions anyway at some point (at least when we fix se-passau/VaRA#626).

vulder replied:
kk, then no

vulder commented on varats/varats/tools/research_tools/research_tool.py:
```diff
@@ -29,6 +30,38 @@
+}
+
+
+class Dependencies:
```
We should include 1-2 tests to check that the install strings are correctly generated.

vulder commented on varats/varats/tools/research_tools/vara.py:
```diff
@@ -110,6 +113,11 @@ class VaRA(ResearchTool[VaRACodeBase]):
     Find the main repo online on github: https://github.com/se-passau/VaRA
     """
 
+    DEPENDENCIES = Dependencies({
+        Distro.DEBIAN: ["libboost-all-dev"],
+        Distro.ARCH: ["boost"]
+    })
```
```suggestion
    __DEPENDENCIES = Dependencies({
        Distro.DEBIAN: ["libboost-all-dev"],
        Distro.ARCH: ["boost"]
    })
```
Should we make this private? Otherwise, users think they can change this.

boehmseb replied:
Not sure. Could be useful to others to query the dependencies of a tool though...

vulder replied:
That's true but then we should provide a get that copies the list. From my current understanding I only see read_only access making sense here.

boehmseb replied:
Would make sense to include that as an abstract class-level function (yes, that is a thing with python's abc) in the base class then.

vulder commented on varats/varats/tools/research_tools/vara.py:
```diff
@@ -256,14 +264,27 @@ def verify_install(self, install_location: Path) -> bool:
+
+        container_vara_dir = containers.CONTAINER_VARATS_ROOT / "tools/VaRA"
+        layers.run(*(self.DEPENDENCIES.get_install_command(distro).split(" ")))
+        layers.copy_([str(self.install_location())], str(container_vara_dir))
```
I think, we still need to handle the copying of shared libraries that are provided via the LD_LIBRARY_PATH

boehmseb replied:
When we use older containers, we might. But for a debian 10 container (the one I tested with) I can simply install boost as a dependency and I don't need to copy anything.
That's why I committed it that way.
Also, I'm not quite sure what's the best way to do this. Maybe look at the BB config and if there is a entry for `LD_LIBRARY_PATH` in the env copy the contents to the container? Or add a mountpoint and always let the `LD_LIBRARY_PATH` point there?

boehmseb replied:
Just copying the thing for now

vulder commented on varats/varats/tools/research_tools/vara.py:
```diff
@@ -110,6 +113,11 @@ class VaRA(ResearchTool[VaRACodeBase]):
     """
 
+    DEPENDENCIES = Dependencies({
+        Distro.DEBIAN: ["libboost-all-dev"],
```
Syntax looks good to me :)

boehmseb committed:
Introduce a context for base container image creation.

boehmseb committed:
Use context to aggregate container env.

This allows to set the same env var in multiple location and append to already defined env vars via BB_ENV

boehmseb committed:
Add tests for Distro

boehmseb commented on varats/varats/containers/containers.py:
```diff
@@ -36,34 +39,146 @@ def prepare_buildah() -> ConcreteCommand:
-CONTAINER_VARATS_ROOT: Path = Path("/varats_root/")
-CONTAINER_BB_ROOT: Path = Path("/app/")
 
+class BaseImageCreationContext():
```
If we feel the need to, we could hide the complete BB-buildah interface behind this context and only allow a couple of "safe" operations. Although only during base image creation. Projects are still allowed to mess everything up if they want to...
Anyway, this should really simplify the research tool integration.

boehmseb committed:
Add check whether distro is in dependencies

boehmseb committed:
Copy LD_LIBRARY_PATH into continer if available

vulder reviewed (approved):
Regarding the context -> sounds good

boehmseb committed:
Extend ResearchTool interface with accessor for dependencies

boehmseb committed:
Merge branch 'f-ContainerSupport' into f-VaraContainerInstall
